File: core/swift/Sources/HeliumPasspoint/ProfileAPIClient.swift
import Foundation

/// Seam that lets the API client be tested without a network.
protocol HTTPTransport {
  func send(_ request: URLRequest) async throws -> (Data, HTTPURLResponse)
}

struct URLSessionTransport: HTTPTransport {
  let session: URLSession

  init(session: URLSession = .shared) {
    self.session = session
  }

  func send(_ request: URLRequest) async throws -> (Data, HTTPURLResponse) {
    let (data, response) = try await session.data(for: request)
    guard let http = response as? HTTPURLResponse else {
      throw PasspointError.apiError("Response was not HTTP")
    }
    return (data, http)
  }
}

/// Talks to the Helium inventory API. Contains no platform-specific code, so it
/// is exercised directly by the unit tests via a stub ``HTTPTransport``.
struct ProfileAPIClient {
  static let apiKeyHeader = "X-Helium-P-Api-Key"
  static let generateProfilePath = "preset/profile/generate"
  static let profileStatusPath = "preset/profile/status"
  static let statusSubscriberQueryParam = "subscriber_id"

  /// `application/x-www-form-urlencoded`-safe set: unreserved characters only,
  /// matching what Java's `URLEncoder` leaves alone (minus `*`, which it keeps
  /// but which is safe to escape).
  static let queryValueAllowed = CharacterSet(
    charactersIn: "ABCDEFGHIJKLMNOPQRSTUVWXYZabcdefghijklmnopqrstuvwxyz0123456789-_.~")

  let baseURL: URL
  let apiKey: String
  let transport: HTTPTransport

  init(baseURL: URL, apiKey: String, transport: HTTPTransport = URLSessionTransport()) {
    self.baseURL = baseURL
    self.apiKey = apiKey
    self.transport = transport
  }

  /// Exchange a CSR for a signed Passpoint profile.
  func generateProfile(
    csr: String,
    subscriberID: String,
    eapType: EAPType,
    presetID: String?
  ) async throws -> IssuedProfile {
    var request = URLRequest(url: baseURL.appendingPathComponent(Self.generateProfilePath))
    request.httpMethod = "POST"
    request.setValue(apiKey, forHTTPHeaderField: Self.apiKeyHeader)
    request.setValue("application/json", forHTTPHeaderField: "Content-Type")
    request.setValue("application/json", forHTTPHeaderField: "Accept")

    var body: [String: Any] = [
      "type": eapType.rawValue,
      "subscriber_id": subscriberID,
      "csr": csr,
    ]
    if let presetID, !presetID.isEmpty {
      body["preset_id"] = presetID
    }
    do {
      request.httpBody = try JSONSerialization.data(withJSONObject: body)
    } catch {
      throw PasspointError.apiError("Failed to encode request body: \(error.localizedDescription)")
    }

    let (data, http) = try await perform(request)
    try Self.throwIfErrorStatus(http.statusCode, data: data)

    do {
      return try JSONDecoder().decode(ProfileResponse.self, from: data).issuedProfile
    } catch {
      throw PasspointError.apiError("Failed to decode profile: \(error.localizedDescription)")
    }
  }

  /// Server-side status for a subscriber, or `nil` when the server has no
  /// profile for them (HTTP 404).
  func profileStatus(subscriberID: String) async throws -> RemoteProfileStatus? {
    // URLQueryItem leaves "+" and "@" unescaped, and a server reading the query
    // string will decode "+" as a space. Percent-encode explicitly so a
    // subscriber ID survives the round trip and matches what the Android SDK
    // (URLEncoder) sends for the same input.
    guard
      let encoded = subscriberID.addingPercentEncoding(
        withAllowedCharacters: Self.queryValueAllowed),
      let url = URL(
        string: baseURL.appendingPathComponent(Self.profileStatusPath).absoluteString
          + "?\(Self.statusSubscriberQueryParam)=\(encoded)")
    else {
      throw PasspointError.apiError("Failed to construct status URL")
    }

    var request = URLRequest(url: url)
    request.httpMethod = "GET"
    request.setValue(apiKey, forHTTPHeaderField: Self.apiKeyHeader)
    request.setValue("application/json", forHTTPHeaderField: "Accept")

    let (data, http) = try await perform(request)
    if http.statusCode == 404 { return nil }
    try Self.throwIfErrorStatus(http.statusCode, data: data)

    do {
      return try JSONDecoder().decode(StatusResponse.self, from: data).remoteStatus()
    } catch {
      throw PasspointError.apiError("Failed to decode status: \(error.localizedDescription)")
    }
  }

  // MARK: - Private

  private func perform(_ request: URLRequest) async throws -> (Data, HTTPURLResponse) {
    do {
      return try await transport.send(request)
    } catch let error as PasspointError {
      throw error
    } catch {
      throw PasspointError.networkError(error.localizedDescription)
    }
  }

  /// Maps HTTP status to error codes exactly as `contract.json` prescribes.
  static func throwIfErrorStatus(_ status: Int, data: Data) throws {
    guard !(200...299).contains(status) else { return }
    let body = String(data: data, encoding: .utf8) ?? ""
    switch status {
    case 401, 403:
      throw PasspointError.apiUnauthorized
    case 429:
      throw PasspointError.apiRateLimited("HTTP \(status)")
    default:
      throw PasspointError.apiError("HTTP \(status): \(body)")
    }
  }
}

File: core/kotlin/src/main/kotlin/com/helium/passpoint/ProfileApiClient.kt
package com.helium.passpoint

import java.net.HttpURLConnection
import java.net.URL
import java.net.URLEncoder
import org.json.JSONArray
import org.json.JSONObject

/** One HTTP request, independent of any particular client library. */
internal data class HttpRequest(
  val url: String,
  val method: String,
  val headers: Map<String, String> = emptyMap(),
  val body: ByteArray? = null,
) {
  // Generated equals/hashCode would compare ByteArray by identity.
  override fun equals(other: Any?): Boolean =
    other is HttpRequest &&
      url == other.url &&
      method == other.method &&
      headers == other.headers &&
      (body?.contentEquals(other.body ?: ByteArray(0)) ?: (other.body == null))

  override fun hashCode(): Int =
    (((url.hashCode() * 31) + method.hashCode()) * 31 + headers.hashCode()) * 31 +
      (body?.contentHashCode() ?: 0)
}

internal data class HttpResponse(val statusCode: Int, val body: String)

/** Seam that lets the API client be tested without a network. */
internal interface HttpTransport {
  /** @throws java.io.IOException on transport failure; the caller maps it to NETWORK_ERROR. */
  fun send(request: HttpRequest): HttpResponse
}

/** Production transport, on the JDK's built-in client so the AAR stays dependency-light. */
internal class HttpUrlConnectionTransport(
  private val connectTimeoutMs: Int = 30_000,
  private val readTimeoutMs: Int = 30_000,
) : HttpTransport {
  override fun send(request: HttpRequest): HttpResponse {
    val connection = (URL(request.url).openConnection() as HttpURLConnection).apply {
      requestMethod = request.method
      connectTimeout = connectTimeoutMs
      readTimeout = readTimeoutMs
      request.headers.forEach { (name, value) -> setRequestProperty(name, value) }
      if (request.body != null) doOutput = true
    }

    try {
      request.body?.let { payload -> connection.outputStream.use { it.write(payload) } }
      val status = connection.responseCode
      val stream = if (status in 200..299) connection.inputStream else connection.errorStream
      val body = stream?.bufferedReader()?.use { it.readText() } ?: ""
      return HttpResponse(status, body)
    } finally {
      connection.disconnect()
    }
  }
}

/**
 * Talks to the Helium inventory API. Contains no Android-specific code, so the
 * unit tests exercise it directly through a fake [HttpTransport].
 */
internal class ProfileApiClient(
  private val baseUrl: String,
  private val apiKey: String,
  private val transport: HttpTransport = HttpUrlConnectionTransport(),
) {
  companion object {
    const val API_KEY_HEADER = "X-Helium-P-Api-Key"
    const val GENERATE_PROFILE_PATH = "preset/profile/generate"
    const val PROFILE_STATUS_PATH = "preset/profile/status"
    const val STATUS_SUBSCRIBER_QUERY_PARAM = "subscriber_id"

    /**
     * RFC 3986 percent-encoding for a query value. [URLEncoder] is
     * form-encoding: it emits `+` for a space, which a server reading the query
     * string decodes back to a space only by convention. Normalising here keeps
     * the URL identical to the one the iOS SDK builds for the same input.
     */
    fun encodeQueryValue(value: String): String =
      URLEncoder.encode(value, "UTF-8")
        .replace("+", "%20")
        .replace("*", "%2A")
        .replace("%7E", "~")

    /** Maps HTTP status to error codes exactly as `contract.json` prescribes. */
    fun throwIfErrorStatus(status: Int, body: String) {
      if (status in 200..299) return
      when (status) {
        401, 403 ->
          throw PasspointException(
            PasspointErrorCode.API_UNAUTHORIZED,
            "API key was rejected (unauthorized).",
          )
        429 ->
          throw PasspointException(
            PasspointErrorCode.API_RATE_LIMITED,
            "Passpoint API rate limit exceeded: HTTP $status",
          )
        else ->
          throw PasspointException(
            PasspointErrorCode.API_ERROR,
            "Passpoint API error: HTTP $status: $body",
          )
      }
    }
  }

  private val root = baseUrl.trimEnd('/')

  /** Exchange a CSR for a signed Passpoint profile. */
  fun generateProfile(
    csr: String,
    subscriberId: String,
    eapType: EapType,
    presetId: String?,
  ): IssuedProfile {
    val payload = JSONObject().apply {
      put("type", eapType.value)
      put("subscriber_id", subscriberId)
      put("csr", csr)
      presetId?.takeIf { it.isNotBlank() }?.let { put("preset_id", it) }
    }

    val response = perform(
      HttpRequest(
        url = "$root/$GENERATE_PROFILE_PATH",
        method = "POST",
        headers = mapOf(
          API_KEY_HEADER to apiKey,
          "Content-Type" to "application/json",
          "Accept" to "application/json",
        ),
        body = payload.toString().toByteArray(Charsets.UTF_8),
      )
    )
    throwIfErrorStatus(response.statusCode, response.body)

    return try {
      val json = JSONObject(response.body)
      IssuedProfile(
        friendlyName = json.getString("friendly_name"),
        domainName = json.getString("domain_name"),
        naiRealmNames = json.getJSONArray("nai_realm_names").toStringList(),
        trustedServerNames = json.getJSONArray("trusted_server_names").toStringList(),
        tlsVersion = json.optString("tls_version", "1.2").ifBlank { "1.2" },
        certificate = json.getString("certificate"),
        caChain = json.getJSONArray("ca_chain").toStringList(),
      )
    } catch (e: PasspointException) {
      throw e
    } catch (e: Exception) {
      throw PasspointException(
        PasspointErrorCode.API_ERROR,
        "Failed to decode profile: ${e.message}",
        e,
      )
    }
  }

  /**
   * Server-side status for a subscriber, or null when the server has no profile
   * for them (HTTP 404).
   */
  fun profileStatus(subscriberId: String): RemoteProfileStatus? {
    val encoded = encodeQueryValue(subscriberId)
    val response = perform(
      HttpRequest(
        url = "$root/$PROFILE_STATUS_PATH?$STATUS_SUBSCRIBER_QUERY_PARAM=$encoded",
        method = "GET",
        headers = mapOf(API_KEY_HEADER to apiKey, "Accept" to "application/json"),
      )
    )
    if (response.statusCode == 404) return null
    throwIfErrorStatus(response.statusCode, response.body)

    return try {
      val json = JSONObject(response.body)
      val rawExpiry = json.getString("expires_at")
      RemoteProfileStatus(
        subscriberId = json.getString("subscriber_id"),
        presetId = json.getString("preset_id"),
        eapType = json.getInt("eap_type"),
        expiresAt = Iso8601.parse(rawExpiry),
        expiresAtRaw = rawExpiry,
        active = json.getBoolean("active"),
      )
    } catch (e: PasspointException) {
      throw e
    } catch (e: Exception) {
      throw PasspointException(
        PasspointErrorCode.API_ERROR,
        "Failed to decode status: ${e.message}",
        e,
      )
    }
  }

  // MARK: - Private

  private fun perform(request: HttpRequest): HttpResponse =
    try {
      transport.send(request)
    } catch (e: PasspointException) {
      throw e
    } catch (e: Exception) {
      throw PasspointException(
        PasspointErrorCode.NETWORK_ERROR,
        "Network error: ${e.message}",
        e,
      )
    }

  // optString, not getString: a null entry in the array yields "" rather than
  // aborting the install with a JSONException.
  private fun JSONArray.toStringList(): List<String> = (0 until length()).map { optString(it) }
}
